x-pack/plugins/enterprise_search/public/applications/shared/flash_messages/flash_messages_provider.tsx

```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React, { useEffect } from 'react';
+import { useValues, useActions } from 'kea';
+import { History } from 'history';
+
+import {
+  FlashMessagesLogic,
+  IFlashMessagesLogicValues,
+  IFlashMessagesLogicActions,
+} from './flash_messages_logic';
+
+interface IFlashMessagesProviderProps {
+  history: History;
+}
+
+export const FlashMessagesProvider: React.FC<IFlashMessagesProviderProps> = ({ history }) => {
+  const { historyListener } = useValues(FlashMessagesLogic) as IFlashMessagesLogicValues;
+  const { listenToHistory } = useActions(FlashMessagesLogic) as IFlashMessagesLogicActions;
+
+  useEffect(() => {
+    if (!historyListener) listenToHistory(history);
+  }, []);
```

Comment:
I think if you made this change, you wouldn't have to store `historyListener` in FlashMessageLogic, which doesn't seem to be used other than to round trip it here.  (and for what its worth, if you are using `historyListener` elsewhere in FlashMessagesLogic, you should pass it in as [props to the logic](https://kea.js.org/docs/api/logic#logic) instead

```
  const { queuedMessages } = useValues(FlashMessagesLogic) as IFlashMessagesLogicValues;
  const { 
    clearFlashMessages,
    clearQueuedMessages,
    setFlashMessages
  } = useActions(FlashMessagesLogic) as IFlashMessagesLogicActions;
  
  const unlisten = history.listen(() => {
    clearFlashMessages();
    setFlashMessages(queuedMessages);
    clearQueuedMessages();
  });
      
  useEffect(() => {
    return () => unlisten(history);
  }, []);
 
```

(not a suggestion because I wrote this in-browser)

Reply:
II think that depends on where we prefer to keep the majority of our logic - do we want it to be in the `FlashMessagesLogic`, or mixed in with with the `<FlashMesagesProvider />` React component file?

To be honest, I do have a moderate preference to keeping it within FlashMessagesLogic. I like that all the state management is in one place and I'm not jumping between files to check how React Router is controlling flashMessages. It also makes testing easier/better, IMO - it means we can test the real FlashMessagesLogic actions and the values that change as a result, without having to either mock its import or mount the component.

Also worth noting is that the concept of storing a listener in Kea state follows the same logic that our HTTP interceptors use:

https://github.com/elastic/kibana/blob/90bd654d7e7b43b64eaf2053c2803d3b44ba6c72/x-pack/plugins/enterprise_search/public/applications/shared/http/http_logic.ts#L54-L57

https://github.com/elastic/kibana/blob/90bd654d7e7b43b64eaf2053c2803d3b44ba6c72/x-pack/plugins/enterprise_search/public/applications/shared/http/http_provider.tsx#L19-L23

which is a pattern I copied from how our current read-only mode interceptor works in ent-search.

Reply:
Okay I understand more. You want FlashMessagesLogic to receive history, and all logic wrt `history` to happen in FlashMessagesLogic.  I still think there's a way to accomplish this where 1) we pass `history` in as a prop to an instanced kea logic instead of using a reducer and 2) the creation of the unlisten callback with `history.listen`, and the calling of that callback happen in `FlashMessagesLogic` `events.beforeMount` and `events.afterMount`.  

I think, even though we are using it in a few places, "calling an action exactly once to set a reducer when a logic is first mounted.  It's the difference between `const spot = new Pet(); spot.setType('dog');` and `const spot = new Pet({type: 'dog'});`. If the `Pet` class required `type` to work, and we were never changed that property of the class after initialization, IMO it makes sense to pass in `type` when creating a new instance of the `Pet` class.  This is a super reductive example, but if I have time later today I'll see if I can make an example PR of how I'd update the read only interceptor, and we can discuss the pro/cons of that approach there

Reply:
I'm going to be honest, I don't fully see the distinction/point you're making here. In your analogy:

`const spot = new Pet(); spot.setType('dog');`
> If the Pet class required type to work [...] it makes sense to pass in type when creating a new instance of the Pet class

To be clear: `FlashMessagesLogic` does not inherently require `history` to work. It requires it to work _the way we are currently used to_, where flash messages reset on page load. However, that behavior was an artifact/byproduct of us having **different FlashMessages** for each view.

The history listener is a shortcut or workaround to wipe messages on navigation, but a generic FlashMessages component by definition does not necessarily need that behavior. We just specifically want it to work that way for us, because that's what we're used to. There's plenty of use cases in general web app development that I can see as perfectly valid where messages should persist on navigation, and views (for example) call `clearFlashMessages` manually.

Reply:
Similarly, `HttpLogic` does not _inherently_ require interceptors to work. HttpLogic has several functions, one of which is storing Kibana's `http` in Kea so that other Logic files can access that lib/value without it having to be manually passed via param. Its other function is adding http interceptors/listeners, and there may be more functions later - we don't know.

The light `<SomethingProvider>` wrappers added for those are a way of initializing/hydrating those logic files with 3rd party deps such as `http` and `history`, but **also** importantly are a way of ensuring that those Kea logic files stay mounted throughout all navigation & interaction within our app.

Hopefully I'm not barking up the wrong tree or over-explaining the wrong thing here. I think it would benefit us to switch to synchronous chat or a call to clear this up, since I'd also like to clarify if you think this is a blocker to getting the PR merged.

Reply:
Closing the loop: Chatted synchronously w/ Byron and this thread isn't a blocker, but we'll follow up later w/ a code example.

Will go ahead and merge this PR now - thanks Byron, and looking forward to continuing this discussion!